### ISSUE TITLE
Unable to publish multiple projects at the same time.

### ISSUE THREAD

**Issue by joelharkes** [dotnet/sdk#7238]
When i publish with `dotnet` cli, the output is:

> publish: Published to ../output/project1
> Published 1/1 projects successfully

Which indicates i could publish multiple projects at te same time (why else say: 1/1)

so i tried this.

## Steps to reproduce

`dotnet publish project1 project2 -o ../output` 

## Expected  behavior

I expect 2 projects to be published in this output folder (each in there own folder, having the same name as the project folder).

## Actual behavior

> Unrecognized command or argument project2

## Environment data

```
.NET Command Line Tools (1.0.0-preview2-1-003155)

Product Information:
 Version:            1.0.0-preview2-1-003155
 Commit SHA-1 hash:  d7b0190bd4

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.14393
 OS Platform: Windows
 RID:         win10-x64
```

**Comment by darioairoldi** [comment]
hello, 
I was able to publish more 2 projects with the following command 

```
- task: DotNetCoreCLI@2
  displayName: dotnet publish backend --configuration $(buildConfiguration)
  name: psPublishReact
  inputs:
    command: publish
    projects: '**/@(01.1 Angular.csproj|02.1 React.csproj)' 
    publishWebProjects: false
    arguments: '--configuration Release --output "$(Build.ArtifactStagingDirectory)/sites/"'
    zipAfterPublish: true

```

according to link:
**File matching patterns reference**
https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/file-matching-patterns?view=azure-devops&viewFallbackFrom=vsts

**Comment by megakid** [comment]
We would love this too, our build times would be greatly reduced if this feature was added 👍

**Comment by Ryan2065** [comment]
This would be helpful for us. 

something like:

```
dotnet publish -o OutDir -PublishAllSeparately
```

```
dotnet publish -o OutDir -ProjectsSeparated Project1,Project2,Project3
```

The it could publish them in folders individually or just publish the ones specified in the folders (sometimes there's libraries in the solution you wouldn't want published by itself)

**Comment by williamdenton** [comment]
found this which seems relevant:

https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-publish
You can publish a solution (not just a project).

```bash
dotnet publish YourSolution.sln -c Release -p:PublishDir=.\publish
```

>If you specify a relative path when publishing a solution, all output for all projects goes into the specified folder relative to the current working directory. To make publish output go to separate folders for each project, specify a relative path by using the msbuild `PublishDir` property instead of the `--output` option. For example, `dotnet publish -p:PublishDir=.\publish` sends publish output for each project to a `publish` folder under the folder that contains the project file.

this was added in .NET Core 3.0

**Comment by StingyJack** [comment]
The earlier build systems for TFS used to dump all the project outputs into a single folder, it was a pain. I had to write a tool to go re-assemble those loose files back into the different programs and library groups and run that at the end of the workflow.  I've authored several hundred definitions in different versions of TFS, AzDo, and other systems and for different organizations and I have yet to find a case where a single out folder was the right choice for building a solution. It really feels like a niche or specific use case that can be covered by a post build script. Can someone who is a proponent of the single folder explain when it is necessary to have this?

**Comment by billybraga** [comment]
We build a single docker image for our Api and Cli projects (we use the Cli project to run k8s jobs, like run migrations before deploying the new Api version). We want to publish those two projects and their dependencies without having the test projects and their dependencies.

**Comment by AVee** [comment]
> The earlier build systems for TFS used to dump all the project outputs into a single folder, it was a pain.

The proposal here is specifically not to dump all output from a solution into a single folder, but to be able to _selectively_ publish multiple projects to a single folder.

> Can someone who is a proponent of the single folder explain when it is necessary to have this?

My usecase is similar to what @billybraga is doing (although there is no docker), I have an API with related CLI applications for tooling. They belong together, share the same libraries and need to be deployed together. This involves things like database migration tools as mentioned above, but in my case also export/export tools intended to be run from shell manually or from cron. 
In Unix-like operating systems it is pretty common to have several related executables sharing supporting libraries which together make up a single package, I don't see a clean way to do this in dotnet.

(Also, it's kinda annoying that `dotnet build` will happily publish all projects to a single folder when presented with a .sln, but claim it's impossible when presented with a list of specific projects)

**Comment by StingyJack** [comment]
> They belong together, share the same libraries and need to be deployed together. This involves things like database migration tools as mentioned above, but in my case also export/export tools intended to be run from shell manually or from cron.

@AVee - Thanks, this makes sense. I've needed this before on rare occasion and would use the post build event on each supporting application to copy its files into the primary app's output folder. Thats probably the cleanest way to do it in dotnet as it behaves the same way locally and on a build server, it's easy to troubleshoot, and as long as you use powershell core commands to do the copying it will work on any platform also. 

> The proposal here is specifically not to dump all output from a solution into a single folder, but to be able to selectively publish multiple projects to a single folder.

I read the ask as wanting to publish two projects with the same build command, each project output goes into a subfolder with the same name as the project, where the subfolders' parent is the publish path. So building a solution with projectA and projectB or naming them individually in the msbuild command along with a publish folder of `c:\temp` should put the output of each in `c::\temp\projectA` and `c:\temp\projectB` respectively.

**Comment by mmalloy-cnb** [comment]
It finally works!

I hope this helps someone else out; it did work for my pipeline, building multiple project with one command.
After trying multiple configurations, this is what worked for me.

I added [modifyOutputPath: true](https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/build/dotnet-core-cli?view=azure-devops#file-structure-for-output-files-is-different-from-previous-builds) to the `dotnet publish` task and it seems to work now. I have not tested with `dotnet build` or other commands.


my YAML task:

``` yaml
- task: DotNetCoreCLI@2
  displayName: 'dotnet publish'
  inputs:
    command: 'publish'
    publishWebProjects: false
    projects: '**/*.csproj'
    arguments: '-p:Version=$(Version) -o $(Build.ArtifactStagingDirectory) --no-restore'
    zipAfterPublish: false
    modifyOutputPath: true    <------ add this
```

**Comment by kbutto** [comment]
^ would be nice to have something like this for Github Actions!

**Comment by jhudsoncedaron** [comment]
We have been publishing with `"dotnet publish tmp.sln -c " + configuration + " -r " + rid + " --self-contained true -p:ShouldUnsetParentConfigurationAndPlatform=false";` however Microsoft has been for some time recommending to not do this via a `.sln` file. There is a `Microsoft.Build.NoTargets` that seems to be able to solve the problem but it does not do so. The resulting error trying to do that is "C:\Program Files\dotnet\sdk\6.0.202\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.ConflictResolution.targets(112,5): error NETSDK1152: Found multiple publish output files with the same relative path: C:\Users\jhudson\dev\formbuilder\src\Cedaron.FormBuilder.Presentation\appsettings.json"

**Comment by mmalloy-cnb** [comment]
> We have been publishing with `"dotnet publish tmp.sln -c " + configuration + " -r " + rid + " --self-contained true -p:ShouldUnsetParentConfigurationAndPlatform=false";` however Microsoft has been for some time recommending to not do this via a `.sln` file. There is a `Microsoft.Build.NoTargets` that seems to be able to solve the problem but it does not do so. The resulting error trying to do that is "C:\Program Files\dotnet\sdk\6.0.202\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.ConflictResolution.targets(112,5): error NETSDK1152: Found multiple publish output files with the same relative path: C:\Users\jhudson\dev\formbuilder\src\Cedaron.FormBuilder.Presentation\appsettings.json"

Ran into this myself, seems Microsoft updated the later SDKs (noticed it when moving to .NET6) to flag this error. I added the `<ErrorOnDuplicatePublishOutputFiles>false</ErrorOnDuplicatePublishOutputFiles>` property to a `<PropertyGroup>` in the .csproj file that was throwing the error and builds are successful.

**Comment by jhudsoncedaron** [comment]
@mmalloy-cnb : I don't think that's the correct solution. I need one `publish` directory for each project.

**Comment by mmalloy-cnb** [comment]
@jhudsoncedaron please refer to my previous [reply](https://github.com/dotnet/sdk/issues/7238#issuecomment-984180075) above. That is what is working for me; where I have a solution containing multiple projects, which are published with one task into an artifact that contains each project output in their own folder. The key was to make sure to have `modifyOutputPath: true` as part of the tasks configuration.

**Comment by aaronr93** [comment]
From [joelharkes](https://github.com/dotnet/sdk/issues/7238#issuecomment-289679835)
> I made a script that builds the dependency trees for all projects and then executes builds in parallel (Promises nodejs) according to the builds. I got a major decrease in build/publish time (on a multi core environment).

We are publishing over a hundred projects that all share a web of 10-20 core dependencies. Publishing in parallel could decimate our deployment time.
@joelharkes this script could be very useful to us. Do you mind sharing the script please?

**Comment by richlander** [comment]
@baronfel

**Comment by joelharkes** [comment]
> From [joelharkes](https://github.com/dotnet/sdk/issues/7238#issuecomment-289679835)
> 
> > I made a script that builds the dependency trees for all projects and then executes builds in parallel (Promises nodejs) according to the builds. I got a major decrease in build/publish time (on a multi core environment).
> 
> @joelharkes this script you mentioned 5 years ago could be very useful to us. Do you mind sharing the script please, if you still have it?
> 
> We are publishing over a hundred projects that all share a web of 10-20 core dependencies. Publishing in parallel could decimate our deployment time.

sorry mate, it was a node js script I think it also only worked with the old `project.json` format which is no longer used.
This was the gist:

1. find all project files
2. make a dependency graph
3. start with the outside/bottom nodes whom have no children/dependendats and trigger build process in separate process. (with forcing not to rebuild dependency projects)
4. every time a build finishes trigger all dependent nodes which only runes once all the dependencies have been finished (either by removing one by one or counting untill you have had them all finished).

### Conclusion

my end result was that in the end it had more or less the same speed as `dotnet build/publish` does a lot of overhead each time a new project is build. that is why i requested this feature.

**Comment by jhudsoncedaron** [comment]
@joelharkes The fastest path is and has always been generate a .sln file and run `publish tmp.sln -c  $CONFIGURATION -r $RID --self-contained true -p:ShouldUnsetParentConfigurationAndPlatform=false`

**Comment by flobernd** [comment]
> @joelharkes The fastest path is and has always been generate a .sln file 

Generating a .slnf from a .sln is a simple solution as well as they are using a json format for these. I linked a python script that handles this task.

**Comment by marcpopMSFT** [comment]
Old issue triage: We do not have plans to support publish for multiple projects at the same time. The recommendation is to use a sln or slnx for that. Note that we don't recommend -o with that though it can work in some situations. Our recommendation is to use https://learn.microsoft.com/en-us/dotnet/core/sdk/artifacts-output

**Comment by jhudsoncedaron** [comment]
@marcpopMSFT : Well I guess that means Microsoft's recommendation against using a sln to publish is overturned.

**Comment by Forgind** [comment]
> [@marcpopMSFT](https://github.com/marcpopMSFT) : Well I guess that means Microsoft's recommendation against using a sln to publish is overturned.

Where did you find that recommendation? I looked for some documentation and found [this](https://learn.microsoft.com/dotnet/core/tools/dotnet-publish#arguments), which seems to explicitly permit publishing .sln files.

**Comment by jhudsoncedaron** [comment]
@Forgind : I wonder if it got overwritten long ago; there's no full text index of the entire history of the docs. Which would of course mean the recommendation is no more.

The only thing left is the dotent build sln bug where it builds an incorrect dependency tree because of the difference between null and empty string in the project properties. There's multiple open issues for this bug; one of them is https://github.com/dotnet/sdk/issues/38575

**Comment by Forgind** [comment]
That issue seems to say that invoking MSBuild (which does happen for dotnet publish) on a .sln will cause problems if some properties change, as, for instance, if you specify --framework from the command line. I don't think that would be supported, but that's a bit different from marcpopMSFT's suggestion. I think trying to build a .sln or .slnx with -o would (sometimes) fail, depending on your project, but if you use the artifacts output path, you can make it work. (Note that in this case, you would not be specifying -o or --framework.)

**Comment by marcpopMSFT** [comment]
To clarify a bit here, publish with sln can work in some circumstances. baronfel believes it's been improved for multi-targeting if all of the dependencies declare the same TFM but that can fall over quickly. So for some situations, it makes sense to use. In more complex scenarios, using a traversal project may be the best bet https://github.com/microsoft/MSBuildSdks/blob/main/src/Traversal/README.md
